File: fullstack-online/unit_07/lectures/post_it/server/src/services/AlbumsService.js
import { dbContext } from "../db/DbContext.js"
import { Forbidden } from "../utils/Errors.js"




class AlbumsService {


  async createAlbum(albumData) {
    const album = await dbContext.Albums.create(albumData)
    await album.populate('creator', 'name picture')
    return album
  }
  async getAllAlbums() {
    const albums = await dbContext.Albums.find().populate('creator', 'name picture')
    return albums
  }
  async getAlbumById(albumId) {
    const album = await dbContext.Albums.findById(albumId).populate('creator', 'name picture')
    if (!album) throw new Error(`No Album with id: ${albumId}`)
    return album
  }

  async archiveAlbum(albumId, userId) {
    const albumToArchive = await this.getAlbumById(albumId)
    if (albumToArchive.creatorId != userId) throw new Forbidden("Could not archive, Improper access to Album")

    albumToArchive.archived = !albumToArchive.archived
    await albumToArchive.save()
    return albumToArchive
  }
}

export const albumsService = new AlbumsService()

File: fullstack-online/unit_09/post_it/server/Services/AlbumsService.cs




namespace post_it.Services;

public class AlbumsService
{
  private readonly AlbumsRepository _repository;

  public AlbumsService(AlbumsRepository repository)
  {
    _repository = repository;
  }

  internal Album ArchiveAlbum(int albumId, Account userInfo)
  {
    Album album = GetAlbumById(albumId);

    if (album.CreatorId != userInfo.Id)
    {
      throw new Exception($"You cannot archive another user's album, {userInfo.Name}!!!");
    }

    album.Archived = !album.Archived;

    _repository.ArchiveAlbum(album);

    return album;
  }

  internal Album CreateAlbum(Album albumData)
  {
    Album album = _repository.CreateAlbum(albumData);
    return album;
  }

  internal Album GetAlbumById(int albumId)
  {
    Album album = _repository.GetAlbumById(albumId);

    if (album == null)
    {
      throw new Exception("Invalid album id: " + albumId);
    }

    return album;
  }

  internal List<Album> GetAlbums()
  {
    List<Album> albums = _repository.GetAlbums();
    return albums;
  }
}